tapdb/assets_store.go

```diff
@@ -883,6 +883,18 @@ func (a *AssetStore) constraintsToDbFilter(
 		assetFilter.AssetIDFilter = assetIDBytes
 		assetFilter.KeyGroupFilter = groupKeyBytes
 
+		// Check if there's exactly one PrevID and set the filters accordingly.
+		if len(query.PrevIDs) == 1 {
```

Comment:
With this alternative approach we wouldn't need to restrict this to a single ID:

```go
--- a/tapdb/assets_store.go
+++ b/tapdb/assets_store.go
@@ -883,18 +883,6 @@ func (a *AssetStore) constraintsToDbFilter(
                assetFilter.AssetIDFilter = assetIDBytes
                assetFilter.KeyGroupFilter = groupKeyBytes
 
-               // Check if there's exactly one PrevID and set the filters accordingly.
-               if len(query.PrevIDs) == 1 {
-                       prevID := query.PrevIDs[0]
-                       encodedOutpoint, err := encodeOutpoint(prevID.OutPoint)
-                       if err == nil {
-                               assetFilter.AnchorPoint = encodedOutpoint
-                               assetFilter.TweakedScriptKey = prevID.ScriptKey.CopyBytes()
-                       }
-                       // If there's an error, we simply skip setting the fields
-                       // TODO, maybe log error or return it?
-               }
-
                // TODO(roasbeef): only want to allow asset ID or other and not
                // both?
 
@@ -2003,7 +1991,48 @@ func (a *AssetStore) ListEligibleCoins(ctx context.Context,
        // have a block height of 0, so we set the minimum block height to 1.
        assetFilter.MinAnchorHeight = sqlInt32(1)
 
-       return a.queryCommitments(ctx, assetFilter)
+       selectedCommitments, err := a.queryCommitments(ctx, assetFilter)
+       if err != nil {
+               return nil, fmt.Errorf("unable to query commitments: %w", err)
+       }
+
+       // If we want to restrict on specific inputs, we do the filtering now.
+       if len(constraints.PrevIDs) > 0 {
+               selectedCommitments = filterCommitmentsByPrevIDs(
+                       selectedCommitments, constraints.PrevIDs,
+               )
+
+               // If this results in an empty list, we return the same error we
+               // would if there were no coins found without the filter.
+               if len(selectedCommitments) == 0 {
+                       return nil, tapfreighter.ErrMatchingAssetsNotFound
+               }
+       }
+
+       return selectedCommitments, nil
+}
+
+// filterCommitmentsByPrevIDs filters the given commitments by the previous IDs
+// given.
+func filterCommitmentsByPrevIDs(commitments []*tapfreighter.AnchoredCommitment,
+       prevIDs []*asset.PrevID) []*tapfreighter.AnchoredCommitment {
+
+       prevIDMatches := func(p *asset.PrevID,
+               c *tapfreighter.AnchoredCommitment) bool {
+
+               return p.OutPoint == c.AnchorPoint && p.ID == c.Asset.ID() &&
+                       p.ScriptKey == asset.ToSerialized(
+                               c.Asset.ScriptKey.PubKey,
+                       )
+       }
+
+       commitmentInList := func(c *tapfreighter.AnchoredCommitment) bool {
+               return fn.Any(prevIDs, func(p *asset.PrevID) bool {
+                       return prevIDMatches(p, c)
+               })
+       }
+
+       return fn.Filter(commitments, commitmentInList)
 }
```

Reply:
love it! Great idea, will implement. thanks for the feedback